### ISSUE TITLE
TypeMap/Converter behavior broken since 2.4.0

### ISSUE THREAD

**Issue by mpaetzold** [modelmapper/modelmapper#608]
The behavior of TypeMap in combination with Converter seems to be broken since 2.4.0. The following test worked well until version 2.4.0. Also, it got unreliable since the order of statements within the `typeMap.addMappings()` block has an impact on the outcome success of the mapping operation.

``` java
package demo;

import org.junit.jupiter.api.Test;
import org.modelmapper.AbstractConverter;
import org.modelmapper.ModelMapper;
import org.modelmapper.TypeMap;

import static org.junit.jupiter.api.Assertions.*;

class ExampleTest {

    AbstractConverter<Location, String> locationToStreetNumberConverter = new AbstractConverter<Location, String>() {
        @Override
        protected String convert(Location source) {
            return (source.getStreetNumber() == null ? "" : source.getStreetNumber())
                    + (source.getStreetNumberSuffix() == null ? "" : source.getStreetNumberSuffix());
        }
    };

    @Test
    public void testOrder1() {
        ModelMapper modelMapper = new ModelMapper();
        modelMapper.getConfiguration().setAmbiguityIgnored(true);

        TypeMap<Location, Address> typeMap = modelMapper.createTypeMap(Location.class, Address.class);
        typeMap.addMappings(mapper -> {
            mapper.map(Location::getStreetName, Address::setStreet);
            mapper.using(locationToStreetNumberConverter).map(src -> src, Address::setHouseNumber);
        });

        Location location = new Location("Example Street", "123", "d");
        Address address = new Address();

        modelMapper.map(location, address);

        assertEquals("Address{street='Example Street', houseNumber='123d'}", address.toString());
    }

    /**
     * Fails since ModelMapper 2.4.0
     *
     * Expected :Address{street='Example Street', houseNumber='123d'}
     * Actual   :Address{street='Location{streetName='Example Street', streetNumber='123', streetNumberSuffix='d'}', houseNumber='123d'}
     */
    @Test
    public void testOrder2() {
        ModelMapper modelMapper = new ModelMapper();
        modelMapper.getConfiguration().setAmbiguityIgnored(true);

        TypeMap<Location, Address> typeMap = modelMapper.createTypeMap(Location.class, Address.class);
        typeMap.addMappings(mapper -> {
            mapper.using(locationToStreetNumberConverter).map(src -> src, Address::setHouseNumber);
            mapper.map(Location::getStreetName, Address::setStreet);
        });

        Location location = new Location("Example Street", "123", "d");
        Address address = new Address();

        modelMapper.map(location, address);

        assertEquals("Address{street='Example Street', houseNumber='123d'}", address.toString());
    }

    class Address {

        private String street;
        private String houseNumber;

        public String getStreet() {
            return street;
        }

        public void setStreet(String street) {
            this.street = street;
        }

        public String getHouseNumber() {
            return houseNumber;
        }

        public void setHouseNumber(String houseNumber) {
            this.houseNumber = houseNumber;
        }

        @Override
        public String toString() {
            return "Address{" +
                    "street='" + street + '\'' +
                    ", houseNumber='" + houseNumber + '\'' +
                    '}';
        }
    }

    class Location {

        private String streetName;
        private String streetNumber;
        private String streetNumberSuffix;

        public Location(String streetName, String streetNumber, String streetNumberSuffix) {
            this.streetName = streetName;
            this.streetNumber = streetNumber;
            this.streetNumberSuffix = streetNumberSuffix;
        }

        public String getStreetName() {
            return streetName;
        }

        public void setStreetName(String streetName) {
            this.streetName = streetName;
        }

        public String getStreetNumber() {
            return streetNumber;
        }

        public void setStreetNumber(String streetNumber) {
            this.streetNumber = streetNumber;
        }

        public String getStreetNumberSuffix() {
            return streetNumberSuffix;
        }

        public void setStreetNumberSuffix(String streetNumberSuffix) {
            this.streetNumberSuffix = streetNumberSuffix;
        }

        @Override
        public String toString() {
            return "Location{" +
                    "streetName='" + streetName + '\'' +
                    ", streetNumber='" + streetNumber + '\'' +
                    ", streetNumberSuffix='" + streetNumberSuffix + '\'' +
                    '}';
        }
    }

}
```

**Comment by chhsiao90** [comment]
Thanks @mpaetzold for reporting this bug with reproducible code. Bug was fixed by #609 and will have a release asap.

**Comment by chhsiao90** [comment]
ModelMapper v2.4.2 released.
